pyomo/core/base/reference.py

```diff
@@ -0,0 +1,372 @@
+    refer to the same :py:class:`Set` objects for all data identifed by
+    the slice, then the resulting indexed component will be indexed by
+    the product of those sets.  However, if all data do not share common
+    set objects, or only a subset of indices in a multidimentional set
+    appear as wildcards, then the resulting indexed component will be
+    indexed by a :py:class:`SetOf` containing a
+    :py:class:`_ReferenceSet` for the slice.
+
+    Parameters
+    ----------
+    reference : :py:class:`_IndexedComponent_slice`
+        component slice that defines the data to include in the
+        Reference component
+
+    ctype : type [optional]
+        the type used to create the resulting indexed component.  If not
+        specified, the data's ctype will be used (if all data share a
+        common ctype).  If multiple data ctypes are found or ctype is
+        ``None``, then :py:class:`IndexedComponent` will be used.
+
+    Examples
+    --------
+
+    .. doctest::
+
+        >>> from pyomo.environ import *
+        >>> m = ConcreteModel()
+        >>> @m.Block([1,2],[3,4])
+        ... def b(b,i,j):
+        ...     b.x = Var(bounds=(i,j))
+        ...
+        >>> m.r1 = Reference(m.b[:,:].x)
+        >>> m.r1.pprint()
+        r1 : Size=4, Index=r1_index
+            Key    : Lower : Value : Upper : Fixed : Stale : Domain
+            (1, 3) :     1 :  None :     3 : False :  True :  Reals
+            (1, 4) :     1 :  None :     4 : False :  True :  Reals
+            (2, 3) :     2 :  None :     3 : False :  True :  Reals
+            (2, 4) :     2 :  None :     4 : False :  True :  Reals
+
+    Reference components may also refer to subsets of the original data:
+
+    .. doctest::
+
+        >>> m.r2 = Reference(m.b[:,3].x)
+        >>> m.r2.pprint()
+        r2 : Size=2, Index=r2_index
+            Key  : Lower : Value : Upper : Fixed : Stale : Domain
+            (1,) :     1 :  None :     3 : False :  True :  Reals
+            (2,) :     2 :  None :     3 : False :  True :  Reals
+
+    Reference components may have wildcards at multiple levels of the
+    model hierarchy:
+
+    .. doctest::
+
+        >>> from pyomo.environ import *
+        >>> m = ConcreteModel()
+        >>> @m.Block([1,2])
+        ... def b(b,i):
+        ...     b.x = Var([3,4], bounds=(i,None))
+        ...
+        >>> m.r3 = Reference(m.b[:].x[:])
+        >>> m.r3.pprint()
+        r3 : Size=4, Index=r3_index
+            Key    : Lower : Value : Upper : Fixed : Stale : Domain
+            (1, 3) :     1 :  None :  None : False :  True :  Reals
+            (1, 4) :     1 :  None :  None : False :  True :  Reals
+            (2, 3) :     2 :  None :  None : False :  True :  Reals
+            (2, 4) :     2 :  None :  None : False :  True :  Reals
+
+    The resulting reference component may be used just like any other
+    component.  Changes to the stored data will be reflected in the
```

Comment:
Does this include using them in Constraints?

Reply:
Yes.  The data object you get back (e.g., by `m.r[2,4]`) *is* the object from the referenced components, so you can use them just like any other component.  Note though that since the Reference must be constructed from a component slice, the component returned by `Reference()` will always be indexed (and therefore `m.r` cannot be used in a constraint).